### PR TITLE
Implementation of new potential skinning mechanism

### PR DESCRIPTION
**Pull request by turt2live**
**Reviewer: Please look at this for sanity. If it looks sane, it'll get merged and replicated to the rest of the project.** It's an open question if we should apply this to *all* components or just the ones riot-web overrides.

**Based upon https://github.com/matrix-org/matrix-react-sdk/pull/3722 - true diff: https://github.com/matrix-org/matrix-react-sdk/compare/pull/3722/head...pull/3723/head**
**Requires https://github.com/vector-im/riot-web/pull/11660**

**Note: This is set to merge against `travis/sourcemaps` because it breaks all the downstream builds.**

Also sorry for the long description, this is basically a science experiment. 

----

With a switch to Only One Webpack™ we need a way to help developers generate the component index without a concurrent watch task. The best way to do this is to have developers import their components, but how do they do that when we support skins? The answer in this commit is to change skinning.

Skinning now expects to receive your list of overrides instead of the react-sdk+branded components. For Riot this means we send over *only* the Vector components and not Vector+react-sdk. 

Components can then be annotated with the `replaceComponent` decorator to have them be skinnable. The decorator must take a string with the dot path of the component because we can't reliably calculate it ourselves, sadly. 

The decorator does a call to `getComponent` which is where the important part of the branded components not including the react-sdk is important: if the branded app includes the react-sdk then the decorator gets executed before the skin has finished loading, leading to all kinds of fun errors. This is also why the skinner lazily loads the react-sdk components to avoid importing them too early, breaking the app.

The decorator will end up receiving null for a component because of the getComponent loop mentioned: the require() call is still in progress when the decorator is called, therefore we can't error out. All usages of getComponent() within the app are safe to not need such an error (the return won't be null, and developers shouldn't use getComponent() after this commit anyways).

The AuthPage, being a prominent component, has been converted to demonstrate this working. Changes to riot-web are required to have this work.

The reskindex script has also been altered to reflect these skinning changes - it no longer should set the react-sdk as a parent. The eventual end goal is to get rid of `getComponent()` entirely as it'll be easily replaced by imports.

### CODE REVIEW

**Comment by t3chguy** [comment]
> The eventual end goal is to get rid of getComponent() entirely as it'll be easily replaced by imports.

+1000

File: .babelrc [hunk 1]
```diff
@@ -1,20 +1,26 @@
...
-            {
-                "globals": ["Error"]
-            }
-        ],
-        "transform-class-properties",
-        "transform-object-rest-spread",
-        "transform-runtime",
-        "add-module-exports",
-        "syntax-dynamic-import"
+        ["@babel/plugin-proposal-decorators", { "legacy": true }],
```
[the remark this answers is not in the document]
Comment (turt2live) [reply .babelrc]: We use legacy decorators elsewhere - it turns out we need this regardless. The modern decorators also aren't compatible with this sort of thing just yet (we need more ES6 code).

File: .buildkite/pipeline.yaml [hunk 2]
```diff
@@ -1,85 +1,111 @@
...
+#      - "echo '--- Installing Dependencies'"
+#      - "./scripts/ci/install-deps.sh"
+#      - "echo '+++ Running Tests'"
+#      - "./scripts/ci/riot-unit-tests.sh"
+#    env:
+#      CHROME_BIN: "/usr/bin/google-chrome-stable"
+#    plugins:
+#      - docker#v3.0.1:
+#          image: "node:10"
+#          propagate-environment: true
```
[the remark this answers is not in the document]
Comment (turt2live) [reply .buildkite/pipeline.yaml]: They are removed by a different PR. Plan is shortly.

File: src/Skinner.js [hunk 3]
```diff
@@ -28,15 +28,31 @@ class Skinner {
...
+
+        // If that failed, check against our own components
         if (!comp) {
-            comp = this.components['views.'+name];
+            // Lazily load our own components because they might end up calling .getComponent()
+            comp = doLookup(require("./component-index").components);
         }
 
+        // Just return nothing instead of erroring - the consumer should be smart enough to
+        // handle this at this point.
```
[the remark this answers is not in the document]
Comment (turt2live) [reply src/Skinner.js]: It could still be a component, but null is just as clear as a component missing in terms of blowing up.

**Comment by turt2live** [comment]
This was pointed at the wrong branch ._.

**Comment by turt2live** [comment]
@dbkr This is only needed on the component to be replaced. The decorator is executed at runtime, and only decorates classes (which may be a problem for some of our codebase, but that bridge hasn't been crossed yet). The lifecycle is roughly:
1. App loads `matrix-react-sdk`'s `index.js`
2. App calls `loadSkin()` with their component-index of undecorated components.
3. The react-sdk caches all of their components and records what they override.
4. The app continues to load, referencing a component from the react-sdk.
5. The react-sdk now finally gets to execute its decorator, which checks against the app's pre-loaded skin and if it finds something it replaces the component to be loaded. If it doesn't find something it just returns the component unaltered. 
6. The app finally gets rendered to the user.

The decorator here is supposed to be used for adding properties/functions to things. One example is adding a decorator for putting an EventEmitter on something (if you had a limitation that prevented you from just extending it normally, like a third party library). Here we replace the entire class instead, which is technically not what's going on: we're actually overriding the component's members with the skinned component's members. This works similar to a replacement because we ultimately override the constructor, lifecycle handlers, and `render` method. 

The end goal is to remove the react-sdk's component index because it doesn't really serve a purpose once we decorate the entire project. The current issue with it is that it needs generating, so though we could remove the watcher and run it manually it's almost certainly going to be forgotten about by developers. If we use imports instead it just magically works and we don't need to worry about "registering" the component in some index somewhere.

**Comment by dbkr** [comment]
Ah right - understood (presumably if you then wanted to override the overriding classes from a third project though, the ones in riot-web would need to be decorated). I would maybe rename `@replaceComponent` as the current name implies it goes on the replacing component. `@replaceableComponent` maybe? Or maybe even just say put it on all components for simplicity, then it can just be `@component`?

> The current issue with it is that it needs generating, so though we could remove the watcher and run it manually it's almost certainly going to be forgotten about by developers

Yep, I get it - just saying that an alternative would be to go back to running reskindex only when we add components and versioning it, then we only have to run it when adding a component. Yeah, people will still forget at that point but would allow us to fix source maps without having to rewrite the skinning stuff, although it looks like this is mostly done now.

Main thing though - could all the excellent explanatory notes and doc in this PR go in tree somewhere, so we don't have to dig this PR out to understand how it works? :)

**Comment by turt2live** [comment]
Renaming it to something helpful can easily be done. `@replaceableComponent` sounds like the least likely to conflict with something weird later on. 

Docs are being written. I have a fear of forgetting how this works myself ._.

Sourcemaps aren't necessarily blocked on this, though it is a massive quality of life improvement (in that we don't have to remember to do things).

File: .eslintrc.js [hunk 4]
```diff
@@ -25,6 +25,7 @@ module.exports = {
     parserOptions: {
         ecmaFeatures: {
             jsx: true,
+            legacyDecorators: true,
```
[the remark this answers is not in the document]
Comment (turt2live) [reply .eslintrc.js]: we have to add it now otherwise eslint complaints aggressively :(

**Comment by turt2live** [comment]
Yup, per the description the plan is to open a subsequent PR that does the conversion for everything else

**Pull request merged**